### PR TITLE
Add delivery report Raredisease

### PR DESCRIPTION
**Pull request by ivadym**
## Description

https://github.com/Clinical-Genomics/cg/pull/3267/files

### Added
- Delivery report Raredisease

### How to prepare for test

- [ ] Ssh to relevant server (depending on type of change)
- [ ] Use stage: `us`
- [ ] Paxa the environment: `paxa`
- [ ] Install on stage (example for Hasta):
    ```shell
    bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-tool-stage.sh -e S_cg -t cg -b [THIS-BRANCH-NAME] -a
    ```

### How to test

- [x] `cg generate delivery-report <raredisease_case_id>`

Integration tests:
- [x] `cg generate delivery-report <balsamic_case_id>`
- [x] `cg generate delivery-report <rnafusion_case_id>`

### Expected test outcome

- [x] Check that RD delivery report is generatedx
- [x] Check that coverage is correctly extarcted from Chanjo2 API
Screenshot 2024-07-26 at 14 25 43

- [x] Check succcesful integration tests
  - [x] Balsamic
  - [x] Rnafusion
  - [ ] MIP-DNA

## Review

- [ ] Tests executed by
- [ ] "Merge and deploy" approved by
Thanks for filling in who performed the code review and the test!

### This [version](https://semver.org/) is a

- [ ] **MAJOR** - when you make incompatible API changes
- [ ] **MINOR** - when you add functionality in a backwards compatible manner
- [ ] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

## Implementation Plan

- [ ] Document in ...
- [ ] Deploy this branch on ...
- [ ] Inform to ...

### CODE REVIEW

File: cg/constants/constants.py [hunk 1]
```diff
@@ -162,10 +162,12 @@ class FileFormat(StrEnum):
 
 
 class GenomeVersion(StrEnum):
+    GRCh37: str = "GRCh37"
+    GRCh38: str = "GRCh38"
+    T2T_CHM13: str = "T2T-CHM13v2.0"
+    canfam3: str = "canfam3"
     hg19: str = "hg19"
     hg38: str = "hg38"
```
Comment (ChrOertlin) [inline cg/constants/constants.py:170]: probably out of scope, but why are these Enums lower case?

File: cg/constants/raredisease.py [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+RAREDISEASE_COVERAGE_THRESHOLD: int = 10
+
+RAREDISEASE_METRIC_CONDITIONS: dict[str, dict[str, Any]] = {
+    "percent_duplicates": {"norm": "lt", "threshold": 20},
+    "PCT_PF_UQ_READS_ALIGNED": {"norm": "gt", "threshold": 0.95},
+    "MEDIAN_TARGET_COVERAGE": {"norm": "gt", "threshold": 26},
+    "PCT_TARGET_BASES_10X": {"norm": "gt", "threshold": 0.95},
+    "PCT_EXC_ADAPTER": {"norm": "lt", "threshold": 0.0005},
+    "predicted_sex_sex_check": {"norm": "eq", "threshold": None},
+}
```
Comment (ChrOertlin) [inline cg/constants/raredisease.py:16]: I think there was a push to put nf related analysis under a single umbrella. @fevac worked on this before I think.

File: cg/meta/report/raredisease.py [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+                case=case, required_fields=REQUIRED_SAMPLE_RAREDISEASE_FIELDS
+            ),
+            "methods": self.get_sample_required_fields(
+                case=case, required_fields=REQUIRED_SAMPLE_METHODS_FIELDS
+            ),
+            "timestamps": self.get_timestamp_required_fields(
+                case=case, required_fields=REQUIRED_SAMPLE_TIMESTAMP_FIELDS
+            ),
+            "metadata": self.get_sample_metadata_required_fields(case=case),
+        }
```
Comment (ChrOertlin) [inline cg/meta/report/raredisease.py:115]: Can we make a pydantic model for this and then return model_to_dict?

RareDiseaseReportRequiredFields(BaseModel) or something

File: cg/meta/workflow/raredisease.py [hunk 4]
```diff
@@ -142,3 +151,46 @@ def get_workflow_metrics(self, sample_id: str) -> dict:
...
+            bundle_name=bundle_name, tags=RAREDISEASE_COVERAGE_FILE_TAGS + [sample_id]
+        )
+        if coverage_file:
+            return coverage_file.full_path
+        LOG.warning(f"No coverage file found with the tags: {RAREDISEASE_COVERAGE_FILE_TAGS}")
+        return None
+
+    def get_gene_ids_from_scout(self, panels: list[str]) -> list[int]:
+        """Return HGNC IDs of genes from specified panels using the Scout API."""
+        gene_ids = []
```
Comment (ChrOertlin) [inline cg/meta/workflow/raredisease.py:167]: typehint

File: cg/meta/workflow/raredisease.py [hunk 5]
```diff
@@ -142,3 +151,46 @@ def get_workflow_metrics(self, sample_id: str) -> dict:
...
+        for panel in panels:
+            genes: list[dict] = self.scout_api.get_genes(panel)
+            gene_ids.extend(
+                gene.get("hgnc_id") for gene in genes if gene.get("hgnc_id") is not None
```
Comment (ChrOertlin) [inline cg/meta/workflow/raredisease.py:171]:
```suggestion
                gene.get("hgnc_id") for gene in genes if gene.get("hgnc_id") is not None
```
"hgnc_id" create constant

File: cg/models/raredisease/raredisease.py [hunk 6]
```diff
@@ -1,20 +1,17 @@
...
-    percentage_mapped_reads: float | None
-    pct_target_bases_10x: float | None
-    median_target_coverage: float | None
-    pct_pf_reads_aligned: float | None
-    pct_pf_reads_improper_pairs: float | None
-    pct_adapter: float | None
-    fraction_duplicates: float | None
+    mapped_reads: int
+    percent_duplicates: float
+    predicted_sex_sex_check: SexOptions
```
Comment (ChrOertlin) [inline cg/models/raredisease/raredisease.py:13]: double check spelling `_sex_sex_`

File: cg/models/report/metadata.py [hunk 7]
```diff
@@ -31,17 +31,35 @@ class MipDNASampleMetadataModel(SampleMetadataModel):
 
     Attributes:
         bait_set: panel bed used for the analysis; source: LIMS
-        gender: gender estimated by the workflow; source: workflow
         mapped_reads: percentage of reads aligned to the reference sequence; source: workflow
         mean_target_coverage: mean coverage of a target region; source: workflow
         pct_10x: percent of targeted bases that are covered to 10X coverage or more; source: workflow
+        sex: sex estimated by the workflow; source: workflow
```
Comment (ChrOertlin) [inline cg/models/report/metadata.py:37]:
```suggestion
        sex: sex predicted by the workflow; source: workflow
```

**Review by ChrOertlin — APPROVED** [review]
Looks good, check with @fevac about pulling the raredisease constants out of the nf_analysis umbrella.

File: cg/constants/constants.py [hunk 8]
```diff
@@ -162,10 +162,12 @@ class FileFormat(StrEnum):
 
 
 class GenomeVersion(StrEnum):
+    GRCh37: str = "GRCh37"
+    GRCh38: str = "GRCh38"
```
Comment (seallard) [inline cg/constants/constants.py:166]: Any chance we can give these descriptive names instead? If there is such a name 🥹 Like `HUMAN_37` or something

File: cg/constants/constants.py (hunk 1 above)
Comment (ivadym) [reply cg/constants/constants.py:170]: no idea, I've changed corrected them

File: cg/models/raredisease/raredisease.py (hunk 6 above)
Comment (ivadym) [reply cg/models/raredisease/raredisease.py:13]: name coming from the pipeline unfortunately :(

File: cg/constants/constants.py (hunk 8 above)
Comment (ivadym) [reply cg/constants/constants.py:166]: The problem is that both terms refer to the same human genome assembly, and we also have more complex names like `T2T_CHM13` for a specific cell line. I would keep the technical notation to keep the names short and consistent with other versions

File: cg/meta/report/raredisease.py (hunk 3 above)
Comment (ivadym) [reply cg/meta/report/raredisease.py:115]: Done

changed it for the rest of the workflows as well

**Comment by ivadym** [comment]
Depoloyed:

Screenshot 2024-07-29 at 09 54 37

**Pull request merged**